VirtusizeCore/Sources/Utils/ExpiringCache.swift

```diff
@@ -0,0 +1,141 @@
+//  copies of the Software, and to permit persons to whom the Software is
+//  furnished to do so, subject to the following conditions:
+//
+//  The above copyright notice and this permission notice shall be included in
+//  all copies or substantial portions of the Software.
+//
+//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+//  THE SOFTWARE.
+//
+
+private struct CacheEntry {
+	let createdAt: Date
+	let value: Any
+	init(_ value: Any) {
+		self.createdAt = Date()
+		self.value = value
+	}
+}
+
+public class ExpiringCache {
+	public static let shared = ExpiringCache()
+
+	private var cache = [String: CacheEntry]()
+
+	/// Remove all cached values
+	public func clearAll() {
+		cache.removeAll()
+	}
+
+	/// Remove value for specified key
+	public func clear(_ key: String) {
+		cache.removeValue(forKey: key)
+	}
+
+	/// Put value into cache by specified key
+	public func set(_ value: Any, forKey key: String) {
+		cache[key] = CacheEntry(value)
+	}
+
+	/// Return cached value if exists, no matter if it's expired or not
+	public func get(_ key: String) -> Any? {
+		guard let cached = cache[key] else {
+			return nil
+		}
+		return cached.value
+	}
+
+	/// Return cached value if it's withing defind TTL (Time To Live) value
+	/// Return "in progress" task, if it's already in progress for the same key
+	/// Fetch, return and cache value, if it's is expired or not yet cached
+	public func getOrFetch(_ key: String, ttl: TimeInterval, fetch: @escaping @Sendable () -> Any) -> Any {
+		if let cached = cache[key] {
+			let lifeTime = Date().timeIntervalSince(cached.createdAt)
+			if lifeTime < ttl {
+				print("> hit cache: .ready for '\(key)' (to live or another \(ttl - lifeTime) sec)")
+				return cached.value
+			} else {
+				print("> hit cache: .expired for '\(key)' (outdated for \(lifeTime - ttl) sec)")
+			}
+		}
+
+		let value = fetch()
+		cache[key] = CacheEntry(value)
+		return value
+	}
+}
+
+extension ExpiringCache {
+	/// Return cached value or load and cache
+	/// - Parameters:
+	///   - key: a string key to lookup and store the value
+	///   - ttl: Time To Live in seconds. If the cached value is too old, the new value is loaded using `fetch` action
+	///   - fetch: a function to execute if no value found or expired. Loaded result is stored into the cache
+	public func getOrFetch<Value>(_ key: String, ttl: TimeInterval, fetch: @escaping @Sendable () -> Value) -> Value {
+		let resultObj: Any = getOrFetch(key, ttl: ttl) { fetch() }
+		return resultObj as! Value // swiftlint:disable:this force_cast
```

Comment:
Is there any better alternative than forcing cast?

Reply:
Alternatively, we can do an explicit error:

```swift
if let value = resultObj as? Value {
    return value
} else {
    fatalError("Type mismatch: Expected \(Value.self), but got \(type(of: resultObj))")
}
```

Reply:
I'm trying to avoid returning nullable type, as it would add extra checks during the usage. 
Do you think making this check explicit on usage is still preferable?

So we would have:

```swift
public func getOrFetch<Value>(...) -> Value? { ... }

func usage() {
  if let value = getOrFetch() {
    ... // use value
  }
}
```

Reply:
```Swift
public func getOrFetch<Value>(_ key: String, ttl: TimeInterval, fetch: @escaping @Sendable () -> Value) -> Value {
    let resultObj: Any = getOrFetch(key, ttl: ttl) { fetch() }
    if let value = resultObj as? Value {
        return value
    } else {
        // Type mismatch, clear the cache and fetch again
        clear(key)
        let newValue = fetch()
        set(newValue, forKey: key)
        return newValue
    }
}
```
What about something like this?

Reply:
Let's rather return `nil` in case of type mismatch. Re-initializing the value looks like "not obvious" behaviour for me.

Reply:
Done